kv/README.md

```diff
@@ -0,0 +1,84 @@
+# Proposed design for JetStream based KV services
+
+This is a proposal to design KV services ontop of JetStream, this requires a KV client to be written, the basic
+features are:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get `string(k)=string(v)` values
+ * Deleting a key completely
+ * Per key TTL
+ * Compacting a key down to n remaining values
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * read-after-write safety unless the read cache of read replicas were used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports 
+ * Key starting with `_kv` is reserved for internal use
+
+This is an implementation of this concept focussed on the CLI, nats.go and others will have to build language
+specific interfaces focussed on performance and end user.
+
+## Design
+
+### Storage
+
+Given a bucket `CONFIGURATION` we will have:
+
+ * A stream called `KV_CONFIGURATION` with subjects `kv.CONFIGURATION.*`
+ * The stream has Max Messages Per Subject limit set to history with optional placement, R and max age for TTL
+ * Getting a value is an ephemeral consumer with filter subject for the key (`kv.CONFIGURATION.foo`), last received in `Direct` mode
+ * A state of `NumPending+Delivered.Consumer` being 0 means the key does not exist
+ * We store headers as per the table below
+
+### Headers
+
+|Header|Description|
+|------|-----------|
+|KV-Origin-Server|The server name where the client was connected to that created the value|
+|KV-Origin-Cluster|The cluster where the client was connected to that created the value|
+|KV-Origin|If the client opts into this, the IP that created the value|
+
+### Watchers
+
+Watchers can either be per key or per the entire bucket.
+
+For watching a key we simply send key updates over the watch, starting with the latest value or nothing. We will only 
+send the last result for a subject - `NumPending==0`.
+
+For watching the bucket we will send a nil if the bucket is empty, else every result even historical ones.
+
+### Read Replicas
+
+A read replicas a mirror stream from the primary stream.  The KV client is configured to do its reads against the 
+named bucket but all writes go to the main bucket for the KV based on above naming.
+
+This will inevitably result in breaking the read-after-write promises and should be made clear to clients.
```

Comment:
can't the client do anything about it? Say cache written results and discard them once an entry with a higher sequence number has been received.

Reply:
Yes, that's in effect how it will work, however there's a delay:

 1. I Put() into CFG
 2. I Get() from CFG_SFO
 3. Stream mirror delivers my write to CFG_SFO

Hence even from a single client the read-after-write is broken. It's probably fine but needs a clear callout.

If you use a local cache then the local cache will alway be built from the main bucket - `CFG` here. Here I do have consistency:

 1. I Put() into the cache, cache invalidates the key and puts into the underlying backend
 2. I Get() on the cache, its a cache miss and fetch from the underlying bucket

OR

 1. I Put() into the cache, cache invalidates the key and puts into the underlying backend
 2. The consumer delivers a new value via a watch and my cache is updated
 3. I Get() on the cache, its a hit and the latest value

Local cache is always correct

Reply:
You could turn it into read my write by caching puts locally until where you read from delivers a higher sequence number than what you stored. 

what I mean is, 
1. I Put(X) into CFG -> get sequence number in return
2. I Get(y) from CFG_SFO -> sequence number is lower, keep x
3. Stream mirror delivers my write to CFG_SFO
2. I Get(z) from CFG_SFO -> sequence number is higher, discard x

Reply:
Unfortunately no, you dont know what the sequence of a put is before its in the stream.  We need to know sequences as some features like consistent puts require a reliable sequence.

This was how I had it at first but then didnt work too hot.

Reply:
so as an example: I am maintaining `_kv/replicas` which has a value like `[bob_sfo, bob_lon]`. I create a new read replica and want to add it to this list, I read the value and then want to put a new one with my replica added.  I do this quite a bit so there's some churn around this key here and perhaps someone else is also writing to it.

To avoid a race I do:

```
val, _ := kv.Get("_kv/replicas")
// do stuff ....
err := kv.Put("_kv/replicas", newValue, OnlyIfLastValueSequence(val.Sequence()))
```

if the put passed I know my calculated data that I put in there was relevant to the most recent key in there and there's no race around other people modifying this key.

If I do too much write-through caching here (updating cache during put) lots of little weirdnesses can creep in, so safest is simply to invalidate the cache for `_kv/replicas` on write